docs(grid): Update cell templating, add CRUD events (#1441)
IgniteUI/igniteui-docfx: en/components/grids_templates/editing.md

```
}

#### Cell Editing Templates

If you want to provide a custom template which will be applied when a cell is in edit mode, you can make use of the [`igxCellEditor` directive]({environment:angularApiUrl}/classes/igxcelltemplatedirective.html). In order to use a custom cell editing template for a column, you need to pass an `ng-template` with the `igxCellEditor` directive and properly bind your custom control to the [`cell.editValue`]({environment:angularApiUrl}/classes/igxcellcomponent.html#editvalue):

```html
<igx-column field="ReorderLevel" header="Reorder Level" [editable]="true">
    <ng-template igxCellEditor let-cell="cell">
        <input type="number" [(ngModel)]="cell.editValue" />
    </ng-template>
</igx-column>
```

> [!NOTE]

> > The cell editing template directive (`igxCellEditor`) is very different from the [cell template(`igxCell`)](../grid/grid.md#cell-template) - the former handles how a column's cells **in edit mode** are displayed an controls the edited cell's **edit value**, the latter control how a column's cell's are show when **outside of edit mode**.

Author: @rkaraivanov , could you take a look when you've got the time?
Does this make a clear enough difference between **cell template** and **cell editing template**, or should I add some extra explanation?

Reviewer: @ViktorSlavov 

IMO, I think the main distinction should be between the functional differences between the two directives. In the end, they both allow for customizing the presentation of cells. Thus an user can still re-template the default cell content through the `igxCell` directive and achieve CRUD functionality even when the column may not be editable. For example:

```html
<ng-template igxCell let-value>
     <!-- This will update the underlying record even when the column is not editable -->
    <input [(ngModel)]="value" /> 
</ng-template>
```

What he won't get however are all the CRUD events, transaction support, etc. I think that should be the focus point of the documentation, backed up by some code snippets/samples which actually show the difference.

```
}

#### Cell Editing Templates

> If you want to provide a custom template which will be applied when a cell is in edit mode, you can make use of the [`igxCellEditor` directive]({environment:angularApiUrl}/classes/igxcelltemplatedirective.html). In order to use a custom cell editing template for a column, you need to pass an `ng-template` with the `igxCellEditor` directive and properly bind your custom control to the [`cell.editValue`]({environment:angularApiUrl}/classes/igxcellcomponent.html#editvalue):

Author: Not exactly, but yeah, I can merge them into one. The first part is "use this when...", the second part - "this is how you use it".

| [`onCellEditCancel`]({environment:angularApiUrl}/classes/igxgridcomponent.html#oncelleditcancel) | Fires when a cell exits edit mode **without committing** its value (e.g. by pressing `Escape`) | [IGridEditEventArgs]({environment:angularApiUrl}/interfaces/igridediteventargs.html) | `true` |
| [`onRowEdit`]({environment:angularApiUrl}/classes/igxgridcomponent.html#onrowedit) | Fires when a row in edit mode's value is **committed** (e.g. by clicking the `Done` button on the Row Editing Overlay) | [IGridEditEventArgs]({environment:angularApiUrl}/interfaces/igridediteventargs.html) | `true` |
| [`onRowEditCancel`]({environment:angularApiUrl}/classes/igxgridcomponent.html#onroweditcancel) | Fires when a row exits edit mode **without committing** its values (e.g. by clicking the `Cancel` button on the Row Editing Overlay) | [IGridEditEventArgs]({environment:angularApiUrl}/interfaces/igridediteventargs.html) | `true` |

All of the above events can be cancelled. For example, if `onCellEditEnter` is cancelled, the cell will never enter edit mode.

> #### Cell validation via edit event cancelling

Reviewer 2: ```suggestion
#### Cell validation on edit event
```
_Not really validation with the cancel, more the other way around. Also the cancelling bit has different significance_

Using the grid's editing events we can alter how the user interacts with the grid.

> In the below example, we'll validate a cell based on the data entered in it by binding to the [`onCellEdit`]({environment:angularApiUrl}/classes/@@igTypeDoc.html#oncelledit) event. If the new value of the cell does not meet our predefined criteria, we'll display a custom error message using [`IgxToast`](../toast.md). We'll also cancel the editing event (by setting `event.cancel = true`), so the user has to specify a valid value before continuing with the editing.

Reviewer 2: Perhaps:
```suggestion
In this example, we'll validate a cell based on the data entered in it by binding to the [`onCellEdit`]({environment:angularApiUrl}/classes/@@igTypeDoc.html#oncelledit) event. If the new value of the cell does not meet our predefined criteria, we'll prevent it from reaching the data source by cancelling the event (`event.cancel = true`). We'll also display a custom error message using [`IgxToast`](../toast.md). 
```
_Swapped those around as "has to specify a valid value before continuing with the editing" isn't always valid, and the important bit from the cancel is that it doesn't commit the data._

Reviewer 2: The tree grid text uses something like "the editing will be cancelled (the value will not be submitted)" which is also fine

            }
        }
    }
}
```
If the value entered in a cell under the **Ordered** column is larger than the available amount (the value under **Units in Stock**), the editing will be cancelled and a toast with an error message will be displayed.
}
@@if (igxName === 'IgxTreeGrid') {
```typescript
export class MyTreeGridEventsComponent {
    ...
    public handleCellEdit(event: IGridEditEventArgs): void {
        const row = this.data.find(e => e[this.grid.primaryKey] === event.cellID.rowID);
        if (!row) {
            return;
        }
        const column = this.grid.columnList.find(col => col.index === event.cellID.columnID);
        if (column.field === "Age") {
            if (event.newValue < 18) {
                event.cancel = true;
                this.toast.message = "Employees must be at least 18 years old!";
                this.toast.show();
            }
        } else if (column.field === "HireDate") {
            if (event.newValue > new Date().getTime()) {
                event.cancel = true;
                this.toast.message = "The employee hire date must be in the past!";
                this.toast.show();
            }
        }
    }
}
```

> Here, we are validating two columns. If the user tries to set an invalid value for an employee's **Age** (below 18) or their hire date (a future date), the editing will be cancelled (the value will not be submitted) and a toast with an error message will be displayed.

Reviewer 2: ```suggestion
Here, we are validating two columns. If the user tries to set an invalid value for an employee's **Age** (below 18) or their **Hire Date** (a future date), the editing will be cancelled (the value will not be submitted) and a toast with an error message will be displayed.
```

}

> @@if (igxName === igxName === 'IgxHierarchicalGrid') {

Reviewer 2: This snippet doesn't display for me. Maybe the extra equals confuses check?
```suggestion
@@if (igxName === 'IgxHierarchicalGrid') {
```

The events can be broken down as follows:

| Event | Description |
|-------|-------------|
| `onCellEditEnter` | Fires when a cell **enters edit mode** |
| `onRowEditEnter` | If `[rowEditing]` is enabled, fires when a row **enters edit mode** |
| `onCellEdit` | Fires when a cell's value is **committed** (e.g. by pressing `Enter`) |
| `onCellEditCancel` | Fires when a cell exits edit mode **without committing** its value (e.g. by pressing `Escape`) |
| `onRowEdit` | Fires when a row in edit mode's value is **committed** (e.g. by clicking the `Done` button on the Row Editing Overlay) |
| `onRowEditCancel` | Fires when a row exits edit mode **without committing** its values (e.g. by clicking the `Cancel` button on the Row Editing Overlay) |

All of the above events can be cancelled. For example, if `onCellEditEnter` is cancelled, the cell will never enter edit mode.

#### Editing Events - Configuration
Using the grid's editing events we can alter how the user interacts with the grid.

> In the below example, we'll validate a cell based on the data entered in it by binding to the [`onCellEdit`]({environment:angularApiUrl}/classes/@@igTypeDoc.html#oncelledit) event. If the new value of the cell does not meet our predefined criteria, we'll display a custom error message (using [`IgxToast`](../toast.md). We'll also cancel the editing event (by setting `event.cancel = true`), so the user has to specify a valid value before continuing with the editing.

Reviewer 2: ```suggestion
In the below example, we'll validate a cell based on the data entered in it by binding to the [`onCellEdit`]({environment:angularApiUrl}/classes/@@igTypeDoc.html#oncelledit) event. If the new value of the cell does not meet our predefined criteria, we'll display a custom error message using [`IgxToast`](../toast.md). We'll also cancel the editing event (by setting `event.cancel = true`), so the user has to specify a valid value before continuing with the editing.
```

The first thing we need to is bind to the grid's event:

```html
<@@igSelector (onCellEdit)="handleCellEdit($event)"
...>
...
</@@igSelector>
```

> @@if (igxName === 'IgxTreeGrid') {

Reviewer 2: _That's Hierarchical Grid config below, no?_

Reviewer 2: Also see https://github.com/IgniteUI/igniteui-angular/issues/5611 - this won't be the case for too long, so I think avoiding this section might actually be better. Plus the `columnList` lookup in the snippet after will only work on the root grid anyways.

Author: Correct, it's for the HGrid, thanks for pointing it out!

```typescript
export class MyHGridEventsComponent {
    ...
    public handleCellEdit(event: IGridEditEventArgs, grid: IgxGridBaseComponent) {
        const today = new Date();
        const column = grid.columnList.find(col => col.index === event.cellID.columnID);
        if (column.field === "Debut") {
            if (event.newValue > today.getFullYear()) {
                this.toast.message = "The debut date must be in the past!";
                this.toast.show();
                event.cancel = true;
            }
        } else if (column.field === "LaunchDate") {
            if (event.newValue > new Date()) {
                this.toast.message = "The launch date must be in the past!";
                this.toast.show();
                event.cancel = true;
            }
        }
    }
}
```
Since we're passing the grid as a second argument in the `handleCellEdit` calls, we're using the passed grid's columns to find to which column the cell belongs. With the above validation in place, when a user tries to change an artist's **Debut** year or an album's **Launch Date**, the grid will not allow any dates that are greater than today.
}

> #### Editing Events - Demo

Reviewer 2: _You don't need to split this section TBH. The entire section can be called "Events Configuration example" or something_

Author: It's convenient if you want to send a link to the demo - a user won't have to scroll through all of the configuration and can just lookup the `StackBlitz` example, in order to see if the implemented functionality suits their needs

> The result of the above validation being applied to our @@igSelector can be seen in the below demo:

Reviewer 2: ```suggestion
The result of the above validation being applied to our `@@igSelector` can be seen in the below demo:
```
